rust/src/ike/logger.rs

```diff
@@ -0,0 +1,227 @@
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::ike::{IKEState, IKETransaction};
+use super::ipsec_parser::IKEV2_FLAG_INITIATOR;
+use crate::ike::parser::{ExchangeType, IsakmpPayloadType, SaAttribute};
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+use std;
+use std::convert::TryFrom;
+use num_traits::FromPrimitive;
+
+const LOG_EXTENDED: u32 = 0x01;
+
+fn add_attributes(transform: &Vec<SaAttribute>, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    for attribute in transform {
+        js.set_string(
+            attribute.attribute_type.to_string().as_str(),
+            format!("{}", attribute.attribute_value.to_string()).as_str(),
+        )?;
+
+        if let Some(numeric_value) = attribute.numeric_value {
+            js.set_uint(
+                format!("{}_raw", attribute.attribute_type).as_str(),
+                numeric_value as u64,
+            )?;
+        } else if let Some(hex_value) = &attribute.hex_value {
+            js.set_string(
+                format!("{}_raw", attribute.attribute_type).as_str(),
+                &hex_value,
+            )?;
+        }
+    }
+
+    return Ok(());
+}
+
+fn log_ike(
+    state: &IKEState, tx: &IKETransaction, flags: u32, jb: &mut JsonBuilder,
+) -> Result<(), JsonError> {
+    jb.open_object("ike")?;
+
+    jb.set_uint("version_major", tx.hdr.maj_ver as u64)?;
+    jb.set_uint("version_minor", tx.hdr.min_ver as u64)?;
+    jb.set_string("init_spi", &tx.hdr.spi_initiator)?;
+    jb.set_string("resp_spi", &tx.hdr.spi_responder)?;
+    jb.set_uint("message_id", tx.hdr.msg_id as u64)?;
+
+    if tx.ike_version == 1 {
+        if let Some(exchange_type) = tx.hdr.ikev1_header.exchange_type {
+            jb.set_uint("exchange_type", exchange_type as u64)?;
+            if (flags & LOG_EXTENDED) == LOG_EXTENDED {
+                if let Some(etype) = ExchangeType::from_u8(exchange_type) {
+                    jb.set_string("exchange_type_verbose", etype.to_string().as_str())?;
+                };
+            }
+        }
+    } else if tx.ike_version == 2 {
+        jb.set_uint("exchange_type", tx.hdr.ikev2_header.exch_type.0 as u64)?;
+    }
+
+    if tx.ike_version == 1 {
+        let mut index = 0;
+        for server_transform in &state.ikev1_container.server.transforms {
+            if index >= 1 {
+                debug_validate_bug_on!(true);
```

Comment:
Hi @satta 
Why is there this assertion ?
It looks like nothing in the code prevents to have multiple transforms, coming from multiple transactions/packets
cf https://redmine.openinfosecfoundation.org/issues/4433

Reply:
Hi, the client is asking the server with a list of proposals, the server is choosing one of them and always sends back that one choice it made. That's the protocol-logic according to the RFC.

Since the packets in both directions have the same format I found it "nicer" to have the same parser logic for both directions and assert afterwards. Perhaps that could even happen earlier while parsing....or it's even something to alert on...not sure here if it can be used in a malicious way.

Unfortunately I have no login to the redmine....

Reply:
Thanks for the quick feedback, I sent you a pcap reproducer by mail.

From your answer, I guess we want to set an event (like `state.set_event(IkeEvent::PayloadExtraData);` but with another event type), instead of using `debug_validate_bug_on`

Reply:
Thanks, I'll have a look!

Yes perhaps is a "softer" logic here better :)

Reply:
Yep, I think an event would be better, too.